methods_exploration/it_functions.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Feb  8 10:11:52 2022
+
+@author: ggorski
+
+These functions were developed borrowing code and ideas from:
+    Laurel Larsen (laurel@berkeley.edu) for Larsen, L. G. and J. W. Harvey. 2017. Disrupted carbon cycling in restored and unrestored urban streams: Critical timescales and controls. Limnology and Oceanography, 62(Suppl. S1), S160-S182, doi: 10.1002/lno.10613. Please cite this work accordingly in products that use this code.
+    Edom Modges (edom.moges@berkeley.edu)
+    Julianne Quinn University of Virginia
+    The transfer entropy routine is based on Ruddell and Kumar, 2009, "Ecohydrologic process networks: 1. Identification," Water Resources Research, 45, doi:10.1029/2008WR007279.
+
+"""
+
+from joblib import Parallel, delayed
+import numpy as np
+
+
+def calc2Dpdf(M,nbins = 11):
+    '''calculates the 3 pdfs, one for x, one for y and a joint pdf for x and y 
+    M: a numpy array of shape (nobs, 2) where nobs is the number of observations
+    this assumes that the data are arrange such that the first column is the source and
+    the second column is the sink.
+    nbins: is the number of bins used for estimating the pdf with a default of 11'''
```

Comment:
Could note that this can be a vector instead if one wanted a different number of bins in x vs. y